rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+
+    IMPORTANT: The ordering of communications must be identical in all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}
+      Replica 1:: `value`: {'a': 3, 'b': [ 2, 98]}
+
+      Replica 0:: result: {'a': 1, 'b': [2, 1]}
+      Replica 1:: result: {'a': 1, 'b': [2, 1]}
+
+    Args:
+      value: The nested structure of `Tensor`s to all-min.
+        The structure must be compatible with `tf.nest`.
+
+    Returns:
+       A `Tensor` nest with the element-wise minimum `value`s from each replica.
+    """
+
+  def all_max(self, value: T) -> T:
+    """All-maxs the given `Tensor` nest across replicas.
+
+    If `all_max` is called in any replica, it must be called in all replicas.
+    The nested structure and `Tensor` shapes must be identical in all replicas.
+
+    IMPORTANT: The ordering of communications must be identical in all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}
+      Replica 1:: `value`: {'a': 3, 'b': [ 2, 98]}
+
+      Replica 0:: result: {'a': 3, 'b': [40, 98]}
+      Replica 1:: result: {'a': 3, 'b': [40, 98]}
+
+    Args:
+      value: The nested structure of `Tensor`s to all-max.
+        The structure must be compatible with `tf.nest`.
+
+    Returns:
+       A `Tensor` nest with the element-wise maximum `value`s from each replica.
+    """
+
+  def all_gather(self, value: T) -> T:
+    """Gathers the given `Tensor` nest from all replicas.
+
+    If `all_gather` is called in any replica, it must be called in all replicas.
+    The nested structure and `Tensor` shapes must be identical in all replicas.
+
+    IMPORTANT: The ordering of communications must be identical in all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}
+      Replica 1:: `value`: {'a': 3, 'b': [ 2, 98]}
+
+      Replica 0:: result: {'a': [1, 3], 'b': [[40, 1], [2, 98]]}
+      Replica 1:: result: {'a': [1, 3], 'b': [[40, 1], [2, 98]]}
+
+    Args:
+      value: The nested structure of `Tensor`s to gather across replicas.
+        The structure must be compatible with `tf.nest`.
+
+    Returns:
+       A `Tensor` nest with the `value`s from each replica. If a `Tensor` within
+       `value` has shape `dims`, the returned `Tensor` will have shape
+       `[num_replicas] + `dims`.
+    """
+
+  def broadcast(self, value: T, source_replica_id: int) -> T:
+    """Broadcasts the given `Tensor` nest from the source to all replicas.
+
+    If `broadcast` is called in any replica, it must be called in all replicas.
+    The nested structure, `Tensor` shapes, and `source_replica_id` must be
+    identical in all replicas.
+
+    IMPORTANT: The ordering of communications must be identical in all replicas.
+
+    `value` will always be evaluated on all replicas.
+
+    Example with two replicas:
+      Replica 0:: `value`: {'a': 1, 'b': [40,  1]}, `source_replica_id`: 1
```

Comment:
source_replica_id: 0 here?

Reply:
`source_replica_id` should be the same on all of them actually, it specifies the replica whose value we want to broadcast.